File: CS425-PA03/src/PlayerControllerComponent.cpp
#include "PlayerControllerComponent.h"
#include "PawnEntity.h"
#include "RoomManager.h"

//world regen func
void Regenerate(Tara::LayerNoRef layer);

const static float MOVEMENT_DISTANCE = 16 * 4; //size of 1 tile;

PlayerControllerComponent::PlayerControllerComponent(Tara::EntityNoRef parent, const std::string& name)
	: Tara::Component(parent, name)
{}

PlayerControllerComponentRef PlayerControllerComponent::Create(Tara::EntityNoRef parent, const std::string & name)
{
	PlayerControllerComponentRef component = std::make_shared<PlayerControllerComponent>(parent, name);
	Register(component);
	return component;
}

void PlayerControllerComponent::OnEvent(Tara::Event& e)
{
	Tara::EventFilter filter(e);
	filter.Call<Tara::KeyPressEvent>(TARA_BIND_FN(PlayerControllerComponent::OnKeyPressEvent));

}

void PlayerControllerComponent::OnUpdate(float deltaTime)
{
	if (!m_SendingToNextLevel){
		auto parent = std::dynamic_pointer_cast<PawnEntity>(GetParent().lock()); //OK becase will be cleaned up and end of func call
		if (parent && !parent->GetTraveling()) {
			auto pos = glm::vec2(parent->GetWorldPosition());
			auto roomPos = RoomManager::WorldCoordToRoomCoord(pos);
			auto centered = RoomManager::IsCentered(pos);
			auto room = RoomManager::Get()->GetRoom(roomPos.x, roomPos.y);
			if (room && centered.first && centered.second && room->GetPerm() == 4 && Tara::PowOfTwo(room->GetDoorState())) {
				//regenerate!
				m_SendingToNextLevel = true;
				Tara::After([](Tara::LayerNoRef layer) {Regenerate(layer); }, 1.0, parent->GetOwningLayer());
			}
		}
	}
}

bool PlayerControllerComponent::OnKeyPressEvent(Tara::KeyPressEvent& e)
{
	auto parent = std::dynamic_pointer_cast<PawnEntity>(GetParent().lock()); //OK becase will be cleaned up and end of func call
	if (!parent) { return false; }

	if (parent->GetTraveling() || !parent->GetAlive()) { return false; }
	auto pos = glm::vec2(parent->GetWorldPosition());
	auto key = e.getKey();
	Tara::Vector dir = { 0,0,0 };
	auto roomPos = RoomManager::WorldCoordToRoomCoord(pos);
	auto centered = RoomManager::IsCentered(pos);
	auto room = RoomManager::Get()->GetRoom(roomPos.x, roomPos.y);

	if (key == TARA_KEY_E) {
		LOG_S(INFO) << "CURRENT ROOM: {" << roomPos.x << "," << roomPos.y << "}";
	}

	if (key == TARA_KEY_S || key == TARA_KEY_DOWN) {
		if (centered.first && (!room || !centered.second || room->GetDoorState() & DOORSTATE_DOWN)) {
			dir.y = -1;
		}
	}
	else if (key == TARA_KEY_W || key == TARA_KEY_UP) {
		if (centered.first && (!room || !centered.second || room->GetDoorState() & DOORSTATE_UP)) {
			dir.y = 1;
		}
	}
	else if (key == TARA_KEY_A || key == TARA_KEY_LEFT) {
		if (centered.second && (!room || !centered.first || room->GetDoorState() & DOORSTATE_LEFT)) {
			dir.x = -1;
		}
	}
	else if (key == TARA_KEY_D || key == TARA_KEY_RIGHT) {
		if (centered.second && (!room || !centered.first || room->GetDoorState() & DOORSTATE_RIGHT)) {
			dir.x = 1;
		}
	}
	else {
		return false;
	}
	if (dir.x == 0 && dir.y == 0) {
		//skip movement
		return true;
	}
	dir *= MOVEMENT_DISTANCE * 4.5; //move 4.5 tiles
	parent->SetTarget(parent->GetWorldPosition() + dir, 0.25 * 4.5);
	//LOG_S(INFO) << "Key pressed event handled: " << (char)key;
	return true;
}

File: CS425-PA03/assets/PlayerControllerComponent.lua
local m_SendingToNextLevel = false
CurrentComponent:ListenForEvents(true)
CurrentComponent:SetOnUpdateCallbackFunction(function()
	if (not m_SendingToNextLevel) then
		local parent = Cast(PawnEntity, CurrentComponent:GetParent())
		if (parent and not parent:GetTraveling()) then
			pos = parent:GetWorldPosition()
			roomPos = RoomManager.WorldCoordToRoomCoord(pos)
			centX, centY = RoomManager.IsCentered(pos)
			room = RoomManager.Get():GetRoom(roomPos.x, roomPos.y)
			if (room and centX and centY and room:GetPerm() == 4 --[[and PowOfTwo(room:GetDoorState())]]) then
				m_SendingToNextLevel = true
				-- regen here
			end
		end
	end
end)

CurrentComponent:SetOnEventCallbackFunction(function(event)
	if (event.Type == "KeyPress") then
		local parent = Cast(PawnEntity, CurrentComponent:GetParent())
		if (not parent) then return false end
		pos = parent:GetWorldPosition()
		key = event.Key
		dir = {x = 0, y = 0}
		roomPos = RoomManager.WorldCoordToRoomCoord(pos)
		centX, centY = RoomManager.IsCentered(pos)
		room = RoomManager.Get():GetRoom(roomPos.x, roomPos.y)
		print(centX)
		print(centY)
		if (key == TARA_KEY_E) then
			print("CURRENT ROOM: {" .. roomPos.x .. "," .. roomPos.y .. "}")
		end
		if (key == TARA_KEY_S or key == TARA_KEY_DOWN) then
			if (centX and (room == nil or not centY or (room:GetDoorState() & 2 > 0))) then
				dir.y = -1
			end
		elseif (key == TARA_KEY_W or key == TARA_KEY_UP) then
			if (centX and (room == nil or not centY or (room:GetDoorState() & 1 > 0))) then
				dir.y = 1
			end
		elseif (key == TARA_KEY_A or key == TARA_KEY_LEFT) then
			if (centY and (room == nil or not centX or (room:GetDoorState() & 4 > 0))) then
				dir.x = -1
			end
		elseif (key == TARA_KEY_S or key == TARA_KEY_RIGHT) then
			if (centY and (room == nil or not centX or (room:GetDoorState() & 8 > 0))) then
				dir.x = 1
			end
		else return false end
		if (dir[1] == 0 and dir[2] == 0) then return true end
		dir.x = dir.x * 64 * 4.5
		dir.y = dir.y * 64 * 4.5
		target = parent:GetWorldPosition()
		target.x = target.x + dir.x
		target.y = target.y + dir.y
		parent:SetTarget(target, 0.25 * 4.5)
		return true
	end
end)